index/segment/fst/reader.go

```diff
@@ -0,0 +1,354 @@
+		return nil, err
+	}
+	defer termsFST.Close()
+
+	pl := r.opts.PostingsListPool.Get()
+
+	iter, iterErr := termsFST.Search(re, minByteKey, maxByteKey)
+	defer iter.Close()
+
+	for {
+		if iterErr == vellum.ErrIteratorDone {
+			break
+		}
+
+		if iterErr != nil {
+			return nil, iterErr
+		}
+
+		_, postingsOffset := iter.Current()
+		nextPl, err := r.retrievePostingsList(postingsOffset)
+		if err != nil {
+			return nil, err
+		}
+		if err := pl.Union(nextPl); err != nil {
+			return nil, err
+		}
+
+		iterErr = iter.Next()
+	}
+
+	return pl, nil
+}
+
+func (r *reader) MatchAll() (postings.MutableList, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) Docs(pl postings.List) (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) AllDocs() (doc.Iterator, error) {
+	return nil, errNotImplemented
+}
+
+func (r *reader) retrievePostingsList(postingsOffset uint64) (postings.List, error) {
+	postingsBytes, err := r.retrieveBytes(r.opts.PostingsData, postingsOffset)
+	if err != nil {
+		return nil, fmt.Errorf("unable to retrieve postings data: %v", err)
+	}
+
+	return fromPilosa(postingsBytes, r.opts.PostingsListPool)
+	// return fromRoaringBytes(postingsBytes, r.opts.PostingsListPool)
+}
+
+func (r *reader) allKeys(fst *vellum.FST) ([][]byte, error) {
+	num := fst.Len()
+	keys := make([][]byte, 0, num)
+
+	iter, iterErr := fst.Iterator(minByteKey, maxByteKey)
+	success := false
+	defer func() {
+		if !success {
+			iter.Close()
+		}
+	}()
+
+	for {
+		if iterErr == vellum.ErrIteratorDone {
+			break
+		}
+		if iterErr != nil {
+			return nil, iterErr
+		}
+		key, _ := iter.Current()
+		keys = append(keys, r.copyBytes(key))
+		iterErr = iter.Next()
+	}
+
+	success = true
+	if err := iter.Close(); err != nil {
+		return nil, err
+	}
+
+	return keys, nil
+}
+
+func (r *reader) retrieveTermsFST(field []byte) (*vellum.FST, error) {
+	termsFSTOffset, exists, err := r.fieldsFST.Get(field)
+	if err != nil {
+		return nil, err
+	}
+
+	if !exists {
+		return nil, fmt.Errorf("no terms known for field: %v", string(field))
+	}
+
+	termsFSTBytes, err := r.retrieveBytes(r.opts.FSTTermsData, termsFSTOffset)
+	if err != nil {
+		return nil, fmt.Errorf("error while decoding terms fst: %v", err)
+	}
+
+	return vellum.Load(termsFSTBytes)
+}
+
+// retrieveBytes assumes the base []byte slice is a collection of (size, payload) pairs,
+// where size is strictly uint64 (i.e. 8 bytes), and next `size` bytes after it are the
+// payload. It assumes `offset` points to the start of one such pair, and retrieves the
+// payload while doing bounds checks to ensure no segfaults.
+func (r *reader) retrieveBytes(base []byte, offset uint64) ([]byte, error) {
+	const sizeofUint64 = 8
+	var (
+		sizeStart = offset
+		sizeEnd   = sizeStart + sizeofUint64
+	)
+	if sizeEnd > uint64(len(base)) {
+		return nil, fmt.Errorf("base bytes too small, length: %d, base-offset: %d", len(base), offset)
+	}
+
+	sizeBytes := base[offset : offset+sizeofUint64]
```

Comment:
This can be `base[sizeStart:sizeEnd` yeah?

Reply:
lol yep. made the change.